lib/api/src/externals/function.rs

```diff
@@ -104,7 +147,7 @@ impl Function {
     pub fn new<FT, F>(store: &Store, ty: FT, func: F) -> Self
     where
         FT: Into<FunctionType>,
-        F: Fn(&[Val]) -> Result<Vec<Val>, RuntimeError> + 'static,
+        F: Fn(&[Val]) -> Result<Vec<Val>, RuntimeError> + 'static + Send + Sync,
```

Comment:
What it means for require a function to be `Send + Sync` ?
Could it still be possible to be called in parallel through different threads?

Reply:
I believe `Sync` with `Fn` does mean that the function can be called from multiple threads at the same time. `Send` just means it can be executed on any thread.

For `Send`, this means functions can't use thread-local variables.

For `Sync`, this means functions should not use unsynchronized global variables.

I can't think of any other cases where this is really an issue right now, I'm sure there are a few other restrictions, but most functions that use `unsafe` correctly or don't use `unsafe` at all should be both `Send` and `Sync`

Reply:
I think it also means that the environment captured by the function must also implement `Send` + `Sync`.

Confirmed with:

```rust
use std::rc::Rc;

fn main() {
    fn check<F>(f: F) where F: Fn() + 'static + Send + Sync {}
    
    let x = Rc::new(42);
    check(|| { let _ = x; });
}
```

which raises the following issue:

```
error[E0277]: `Rc<i32>` cannot be shared between threads safely
 --> src/main.rs:7:5
  |
4 |     fn check<F>(f: F) where F: Fn() + 'static + Send + Sync {}
  |                                                 ---- required by this bound in `check`
...
7 |     check(|| { let _ = x; });
  |     ^^^^^ `Rc<i32>` cannot be shared between threads safely
  |
  = help: the trait `Sync` is not implemented for `Rc<i32>`
  = note: required because of the requirements on the impl of `Send` for `&Rc<i32>`
  = note: required because it appears within the type `[closure@src/main.rs:7:11: 7:28]`
```

([see the playground](https://play.rust-lang.org/))

Reply:
@Hywan , I think having a `Send + Sync` environment is ultimately what we need for the Wasmer API... because `Instance` is `Send + Sync`, so a single function can be called from multiple threads concurrently and in parallel.  I don't know a good way around this yet but I think `Send + Sync` does fix everything for us

Reply:
The alternative might be to have separate code and abstractions for non-thread safe things, but that seems pretty complicated 🤔 . I think encouraging code to work on multiple threads by default is a very Rusty thing anyways and I think it positions us and users of `wasmer` well to scale their applications